src/Grpc.Net.Client/Internal/UserAgentGenerator.cs

```diff
@@ -0,0 +1,164 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Runtime.Versioning;
+
+namespace Grpc.Net.Client.Internal
+{
+    internal static class UserAgentGenerator
+    {
+        internal static string GetUserAgentString()
+        {
+            var assembly = typeof(GrpcProtocolConstants).Assembly;
+
+            var assemblyVersion = assembly
+                .GetCustomAttributes<AssemblyInformationalVersionAttribute>()
+                .FirstOrDefault()?
+                .InformationalVersion;
+
+            var frameworkName = assembly
+                .GetCustomAttributes<TargetFrameworkAttribute>()
+                .FirstOrDefault()?
+                .FrameworkName;
+            
+            return GetUserAgentString(
+                processArch: RuntimeInformation.ProcessArchitecture,
+                runtimeVersion: IsMono() ? null : Environment.Version, 
+                assemblyVersion: assemblyVersion,
+                // RuntimeInformation.FrameworkDescription is only supported for
+                // .NET Framework 4.7.1 and above. If targetting net461 or earlier,
+                // the framework description will need to be resolved manually
+                // using reflection.
+                runtimeInformation: RuntimeInformation.FrameworkDescription,
+                frameworkName: frameworkName,
+                operatingSystem: GetOS());
+
+            bool IsMono() => Type.GetType ("Mono.Runtime") != null;
+        }
+
+
+        // Factored out for testing
+        internal static string GetUserAgentString(Architecture processArch, Version? runtimeVersion, string? assemblyVersion, string? runtimeInformation, string? frameworkName, string? operatingSystem)
+        {
+            var userAgent = "grpc-dotnet";
+
+            // /2.41.0-dev
+            userAgent += $"{GetGrpcDotNetVersion(assemblyVersion)} ";
+            // (.NET 5.0.7;
+            userAgent += $"({GetFrameworkDescription(runtimeInformation)}";
+            // CLR 5.0.0;
+            userAgent += GetRuntimeVersion(runtimeVersion);
+            // net6.0;
+            userAgent += GetFrameworkName(frameworkName);
+            // windows  
+            userAgent += $"{operatingSystem} ";
+            // x64)
+            userAgent += $"{GetProcessArch(processArch)})";
+
+            string GetRuntimeVersion(Version? runtimeVersion) => runtimeVersion != null ? $"CLR {runtimeVersion}; " : string.Empty;
```

Comment:
Avoid possibility of accidental closure
```suggestion
            static string GetRuntimeVersion(Version? runtimeVersion) => runtimeVersion != null ? $"CLR {runtimeVersion}; " : string.Empty;
```